src/table_mgr.cc

```diff
@@ -252,6 +246,57 @@ Status TableMgr::init(const TableMgrOptions& _options) {
             }
         }
 
+        // Adjust num L0 partitions blockingly only when it is not logSectionOnly mdoe,
+        // not read only mode, and the number of L0 partitions read from existing manifest
+        // file is different from the number specified in db config.
+        if (!db_config->logSectionOnly && !db_config->readOnly
+            && numL0Partitions != db_config->numL0Partitions) {
+            _log_info(myLog,
+                      "adjust numL0 partitions: %zu -> %zu",
+                      numL0Partitions,
+                      db_config->numL0Partitions);
+
+            if (!db_config->nextLevelExtension) {
+                _log_err(myLog, "[Adjust numL0] not allowed in L0 only mode");
+                throw Status(Status::INVALID_CONFIG);
+            }
+
+            // Need to compact all existing L0 tables to L1 and recreate empty L0 tables,
+            // otherwise hash will be messed up.
+            for (size_t ii = 0; ii < numL0Partitions; ++ii) {
+                // Force compact L0 table to L1 in blocking manner to reduce L0
+                // partitions.
+                std::list<TableInfo*> tables;
+                s = mani->getL0Tables(ii, tables);
+                if (tables.size() != 1 || !s) {
+                    _log_err(myLog, "[Adjust numL0] tables of hash %zu not found", ii);
+                    throw s;
+                }
+                s = compactLevelItr(CompactOptions(), tables.back(), 0, true);
+                if (!s) {
+                    _log_err(myLog, "[Adjust numL0] compaction error: %d", s);
+                    throw s;
+                }
+                // The compacted table is remove from manifest in compactLevelItr,
+                // just release
+                for (TableInfo*& table: tables) {
+                    table->done();
+                }
```

Comment:
This is not that simple problem. What if the process is force-closed in the middle so that a few tables were removed and the others are not, and then we reopen the DB? L0 is screwed up and how are you going to continue adjusting L0? Data loss or letting DB instance unavailable is not acceptable.